Tracked addition of federation, sed support in kube DNS

MikeSpreitzer opened the pull request:
[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

The kube DNS app recently gained support for federation (whatever that
is), including a new Salt parameter.  This broke the deployAddons.sh script for cluster ubuntu.  The DNS app also gained alternate
templates, intended to be friendly to `sed`.  Fortunately, those do
not demand a federation parameter.

This PR fixes up the `cluster/ubuntu/deployAddons.sh` script to track those changes, by switching to the `sed`-friendly templates.

MikeSpreitzer committed:
Tracked addition of federation, sed support in kube DNS

The kube DNS app recently gained support for federation (whatever that
is), including a new Salt parameter.  It also gained alternate
templates, intended to be friendly to `sed`.  Fortunately, those do
not demand a federation parameter.

MikeSpreitzer commented:
This is a bug fix, no release note needed.

roberthbailey commented:
/cc @quinton-hoole FYI (since this broke due to federation changes).

roberthbailey commented:
This looks like a cherry pick candidate because otherwise this will be broken on the 1.3 release branch.

quinton-hoole commented:
@madhusudancs Could you take a look at this please?

madhusudancs commented on cluster/ubuntu/deployAddons.sh:
```diff
@@ -41,8 +41,8 @@ function init {
-  sed -e "s/{{ pillar\['dns_replicas'\] }}/${DNS_REPLICAS}/g;s/{{ pillar\['dns_domain'\] }}/${DNS_DOMAIN}/g;" "${KUBE_ROOT}/cluster/saltbase/salt/kube-dns/skydns-rc.yaml.in" > skydns-rc.yaml
-  sed -e "s/{{ pillar\['dns_server'\] }}/${DNS_SERVER_IP}/g" "${KUBE_ROOT}/cluster/saltbase/salt/kube-dns/skydns-svc.yaml.in" > skydns-svc.yaml
+  sed -e "s/\\\$DNS_REPLICAS/${DNS_REPLICAS}/g;s/\\\$DNS_DOMAIN/${DNS_DOMAIN}/g;" "${KUBE_ROOT}/cluster/saltbase/salt/kube-dns/skydns-rc.yaml.sed" > skydns-rc.yaml
+  sed -e "s/\\\$DNS_SERVER_IP/${DNS_SERVER_IP}/g" "${KUBE_ROOT}/cluster/saltbase/salt/kube-dns/skydns-svc.yaml.in" > skydns-svc.yaml
```
Should you also not switch to the .sed script here for consistency?

roberthbailey replied:
I think so. The first time I looked I didn't see the other .sed file but on second look it's right there.

madhusudancs commented:
Minor nit, LGTM otherwise.

girishkalele commented:
@MikeSpreitzer 

The skydns-rc.yaml.base file is transformed into the .in and .sed templates so that its easy for changes to the base file to propagate to various formats.

Any changes to the .sed templates might break the Ubuntu case again, so for people making changes to the DNS templates to have visibility into your consumption of this template, one suggestion is to add your own skydns-rc.yaml.ubuntu generated file and transforms2ubuntu.sed script and add a Makefile target. That would make it a visible file in this directory.

MikeSpreitzer commented:
(1) Oops, I see I forgot to edit the file extension for the second substitution; I will fix that.

(2) I am surprised by the suggestion to change the Makefile to refer specifically to the cluster/ubuntu/deployAddons.sh script.  I did a little search for "skydns-rc.yaml", and found many other files that also seem to know the required set of substitutions.  The beauty of the sed files is that they can be used by multiple scripts.  It would seem tedious and odd to make special template files for each script.  Perhaps instead we should add a comment somewhere reminding developers to do that search for scripts that use the templates?

MikeSpreitzer committed:
Fixed deployAddons.sh to use sed template for skydns-svc.yaml

My original change neglected to change the template from the salt one
to the sed one.

MikeSpreitzer committed:
Added note to developers about adding parameters to skydns

Added a friendly note, with a suggestion of how to find the scripts.

madhusudancs commented:
LGTM for now.

madhusudancs commented:
@k8s-bot test this issue: #IGNORE
